src/interface/packages.hpp

```diff
@@ -26,15 +27,61 @@ class Packages_t {
   Packages_t() = default;
   void Add(const std::shared_ptr<StateDescriptor> &package);
 
-  std::shared_ptr<StateDescriptor> const &Get(const std::string &name) {
+  std::shared_ptr<StateDescriptor> const &Get(const std::string &name) const {
     return packages_.at(name);
   }
 
+  // Templated version for retrieving a package with a particular type
+  // Allows subclassing 'StateDescriptor' to add user package types to list
+  template<typename T>
+  T* const &Get(const std::string &name) const {
+    return static_cast<T*>(packages_.at(name).get());
+  }
+
   const Dictionary<std::shared_ptr<StateDescriptor>> &AllPackages() const {
     return packages_;
   }
   Dictionary<std::shared_ptr<StateDescriptor>> &AllPackages() { return packages_; }
 
+  // Returns a sub-Dictionary containing just pointers to packages of type T.
+  // Dictionary is a *new copy*, and members are bare pointers, not shared_ptr.
+  template <typename T>
+  const std::vector<T*> AllPackagesOfType() const {
+    Dictionary<T*> sub_dict;
+    for (auto package : packages_) {
+      if (T *cast_package = dynamic_cast<T*>(package.second.get())) {
+        sub_dict[package.first] = cast_package;
+      }
+    }
+    return sub_dict;
+  }
+
+  // Returns a list of pointers to packages of type T.
+  // List contains bare pointers, not shared_ptr objects
+  template <typename T>
+  const std::vector<T*> ListPackagesOfType() const {
+    std::vector<T*> sub_list;
+    for (auto package : packages_) {
+      if (T *cast_package = dynamic_cast<T*>(package.second.get())) {
+        sub_list.append(cast_package);
+      }
+    }
+    return sub_list;
+  }
```

Comment:
I considered that, but I think it would be common/useful to add other elements besides just the function call -- e.g. all my loops right now include extra code that can print the function being called, and I can see them as a good point for adding other debugging or common code.

I also sometimes (often) need control of the order that functions are called, e.g. one "floors" function relies on the results of previous.  Currently I just hack it by pulling out the package I know needs to be first, then looping over the others.  A more general approach might be callback "priorities," but that gets hard to follow reading the code and is more complex to implement so it's not clear we'd want it.

Since only maybe half my callers look anything like that prototype, and I gave up trying to cover the different cases.  I'm happy to provide it for other implementations though.

Reply:
Nah it shouldn't matter much.